core/templates/domain/objects/NumberWithUnitsObjectFactorySpec.ts

```diff
@@ -304,5 +304,23 @@ describe('NumberWithUnitsObjectFactory', () => {
       )).toMathjsCompatibleString())
         .toThrowError('Number with type fraction cannot have a real part.');
     });
+
+    it('should convert units to their canonical form', () => {
+      expect(new NumberWithUnits('real', 1, new Fraction(
+        false, 0, 0, 1), new Units(
+        [{unit: 'newton', exponent: 1}, {unit: 'meter', exponent: -1},
+          {unit: 'celsius', exponent: -2}]))
+        .getCanonicalRepresentationOfUnits()).toEqual(new Units(
+        [{unit: 'N', exponent: 1}, {unit: 'm', exponent: -1},
+          {unit: 'degC', exponent: -2}]).units);
+
+      expect(new NumberWithUnits('real', 1, new Fraction(
+        false, 0, 0, 1), new Units(
+        [{unit: 'rupee', exponent: 1}, {unit: 'dollars', exponent: -1},
+          {unit: '$', exponent: -2}]))
+        .getCanonicalRepresentationOfUnits()).toEqual(new Units(
```

Comment:
This is an interesting case -- shouldn't the canonical representation combine the two "dollars", and probably also sort the unit list by the unit name, so that there really is one canonical representation?

Either that, or we need to raise an error in the duplicate-unit case and tell the user it's invalid if they have two units that mean exactly the same thing (which might actually be better). (We should probably still sort the units during normalization though, since technically kg m and m kg are the same thing.)

Reply:
The current functionality is that we allow duplicate units, and mathjs handles the combination part. for example, if we enter "kg^2/kg", we get "kg". [Here](https://github.com/oppia/oppia/blob/a398151f3ec6d818a487561213043602f92a6112/core/templates/domain/objects/NumberWithUnitsObjectFactorySpec.ts#L41) we have a test for the same. I made a mistake here, my test should contain only one 'dollar' key.

do we need to continue with the current functionality, allowing duplicate units, or should we disallow them?

Reply:
@AFZL210 I think it's OK to (a) do the collapsing of duplicate units here (and sorting if needed) as part of figuring out the canonical representation, and (b) file an issue to catch cases with duplicate units and show the user a warning, if we don't do that already.

Reply:
Sure, I will update my tests. and will create an issue too.

Reply:
(a) mathjs handles the collapsing part already, and we don't really need sorting.

(b) although the change is not big, we just need to add a check in the 'fromRawInputString' function and update some tests. but it's better to have an issue for it to keep track of it (which I will add right after this PR)

Reply:
@AFZL210 For (a), a canonical representation should be canonical and unique, and not reliant on third-party libs which might use the data. Is your canonical representation unique? If not, please fix it.

For (b), please file the issue and link it here, thanks.

Reply:
Yes, confirmed the units.

Reply:
@AFZL210 You haven't answered either parts (a) or (b) of my question.

Reply:
(a) Yes, the canonical representations are unique, and does not depend on anything.
(b) Filed the issue #20044